apps/extension/src/background/approvals/service.ts

```diff
@@ -46,12 +47,81 @@ export class ApprovalsService {
 
   constructor(
     protected readonly txStore: KVStore<TxStore>,
+    protected readonly dataStore: KVStore<string>,
     protected readonly connectedTabsStore: KVStore<TabStore[]>,
     protected readonly approvedOriginsStore: KVStore<ApprovedOriginsStore>,
     protected readonly keyRingService: KeyRingService,
     protected readonly ledgerService: LedgerService,
     protected readonly vaultService: VaultService
-  ) { }
+  ) {}
+
+  async approveSignature(
+    signer: string,
+    data: string
+  ): Promise<SignatureResponse> {
+    const msgId = uuid();
+
+    await this.dataStore.set(msgId, data);
+    const baseUrl = `${browser.runtime.getURL(
+      "approvals.html"
+    )}#/approve-signature/${signer}`;
+
+    const url = paramsToUrl(baseUrl, {
+      msgId,
+    });
+    const approvalWindow = await this._launchApprovalWindow(url);
+    const popupTabId = approvalWindow.tabs?.[0]?.id;
+
+    if (!popupTabId) {
+      throw new Error("no popup tab ID");
+    }
+
+    if (popupTabId in this.resolverMap) {
+      throw new Error(`tab ID ${popupTabId} already exists in promise map`);
+    }
+
+    return await new Promise((resolve, reject) => {
+      this.resolverMap[popupTabId] = { resolve, reject };
+    });
+  }
+
+  async submitSignature(
+    popupTabId: number,
+    msgId: string,
+    signer: string
+  ): Promise<void> {
+    const data = await this.dataStore.get(msgId);
+    const resolvers = this.resolverMap[popupTabId];
+
+    if (!resolvers) {
+      throw new Error(`no resolvers found for tab ID ${popupTabId}`);
+    }
+
+    if (!data) {
+      throw new Error(`Signing data for ${msgId} not found!`);
+    }
+
+    try {
+      const signature = await this.keyRingService.signArbitrary(signer, data);
+      resolvers.resolve(signature);
+    } catch (e) {
+      resolvers.reject(e);
+    }
+    resolvers.resolve();
```

Comment:
I think this won't do anything since the promise will already be resolved/rejected before this line.